packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,398 @@
+) {
+    const { observedAttributes, attributeChangedCallback } = instancedDefinition;
+    if (observedAttributes.size === 0 || isUndefined(attributeChangedCallback)) {
+        return;
+    }
+    const offset = getObservedAttributesOffset(registeredDefinition, instancedDefinition);
+    if (offset.size === 0) {
+        return;
+    }
+    // instance level patches
+    defineProperties(instance, {
+        setAttribute: {
+            value: function setAttribute(name: string, value: any) {
+                if (offset.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    // maybe we want to call the super.setAttribute rather than the native one
+                    nativeSetAttribute.call(this, name, value);
+                    attributeChangedCallback!.call(this, name, old, value + '');
+                } else {
+                    nativeSetAttribute.call(this, name, value);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+        removeAttribute: {
+            value: function removeAttribute(name: string) {
+                if (offset.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    // maybe we want to call the super.removeAttribute rather than the native one
+                    nativeRemoveAttribute.call(this, name);
+                    attributeChangedCallback!.call(this, name, old, null);
+                } else {
+                    nativeRemoveAttribute.call(this, name);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+    });
+}
+
+// User extends this HTMLElement, which returns the CE being upgraded
+let upgradingInstance: HTMLElement | undefined;
+
+// Helper to upgrade an instance with a CE definition using "constructor call trick"
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instancedDefinition: Definition
+) {
+    setPrototypeOf(instance, instancedDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instancedDefinition);
+    // attributes patches when needed
+    if (instancedDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instancedDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    try {
+        new instancedDefinition.UserCtor();
+    } catch (err) {
+        if (err instanceof TypeError && err.message === 'Illegal constructor') {
+            new instancedDefinition.UserCtor();
```

Comment:
this is definitely wrong.

Reply:
Yeah this line is not covered in our tests, let me try removing the `try`/`catch` entirely.